Purges and deletes feedback docs after replication

dianabarsan opened the pull request:
# Description

Purges `feedback` and `telemetry`docs from the user's remote meta database after replicating to `users-meta` database. Purging a document removes it from the `_changes` feed, which insures that no new logins would download old feedback or telemetry docs. 
To handle older replicated documents, executes a "one-time" purge of all feedback docs that have been replicated in the past.

Deletes replicated `feedback` and `telemetry` docs from the user's local meta database when bootstrapping, limiting to deleting 1000 documents at a time. 

medic/cht-core#5902

# Code review items

- Readable: Concise, well named, follows the [style guide](https://github.com/medic/medic-docs/blob/master/development/style-guide.md), documented if necessary.
- Documented: Configuration and user documentation on [medic-docs](https://github.com/medic/medic-docs/)
- Tested: Unit and/or e2e where appropriate
- Internationalised: All user facing text
- Backwards compatible: Works with existing data and configuration or includes a migration. Any breaking changes documented in Changes.md.

# License

The software is provided under AGPL-3.0. Contributions to this project are accepted under the same license.

dianabarsan commented:
@garethbowen please have a look. Thanks!

garethbowen commented on webapp/src/js/services/db-sync.js:
```diff
@@ -4,7 +4,8 @@ const DDOC_PREFIX = ['_design/'];
 const SYNC_INTERVAL = 5 * 60 * 1000; // 5 minutes
-const META_SYNC_INTERVAL = 30 * 60 * 1000; // 30 minutes
+//const META_SYNC_INTERVAL = 30 * 60 * 1000; // 30 minutes
+const META_SYNC_INTERVAL = 30  * 1000; // 30 seconds
```
Oops!

dianabarsan replied:
Whoops. Serves me right in rushing this to get reviewed :(

garethbowen commented on webapp/src/js/services/db-sync.js:
```diff
@@ -157,7 +158,16 @@ angular
+              .map(doc => ({ _id: doc._id, _rev: doc._rev, _deleted: true, purged: true }));
+            return local.bulkDocs(deletes);
+          }
+        });
```
My only concern here is that we won't clean up docs that were created before upgrading to 3.9.0. I'm wondering if we should record the current seq when we start and then, after the sync succeeds, purge everything up to that seq?

dianabarsan replied:
That works! Esp because we can actually compare seqs in PouchDB :) 
I just pushed the test-less code, I'll get back to adding unit tests once the high prio tasks are taken care of.

dianabarsan commented on sentinel/src/schedule/replications.js:
```diff
@@ -37,11 +43,102 @@ function getSchedule(config) {
+const batchedPurge = (sourceDb, targetDb, lastSeq = 0) => {
+  const opts = {
+    since: lastSeq,
+    limit: BATCH_SIZE,
```
Should we do two BATCH_SIZES? One for getting changes (likely larger) and then mini-batch the purges?

garethbowen replied:
I think it's ok. It'd be a little more efficient to get a larger batch, but the code would be more complex.

dianabarsan commented on webapp/src/js/bootstrapper/translator.js:
```diff
@@ -8,6 +8,7 @@ const TRANSLATIONS = {
     LOAD_APP: 'Loading app…',
     PURGE_INIT: 'Checking data…',
     PURGE_INFO: ({ count }) => `Cleaned ${count} documents…`,
+    PURGE_META: 'Cleaning meta data…',
```
Suggestions?

garethbowen replied:
Yeah that's ok. I can't think of a way to explain it that the user's will understand. Leave it for now...

twisteddiana commented:
I've been trying to get translations, but so far not so good progress. So this is still pending.

dianabarsan commented:
I've been trying to get translations, but so far have not had good progress, so still pending.

dianabarsan commented:
@garethbowen I pushed a commit that fixes a bug where the user meta ddoc wasn't being downloaded locally. Could you, please, re-review?

dianabarsan commented:
Hey @garethbowen 
There's been another turn, a more serious one this time! Details are in this comment: https://github.com/medic/cht-core/issues/5902#issuecomment-623368269

I pushed some code that adds a `validate_doc_update` to user metas. Please tell me what you think. 
Thanks for your time!

dianabarsan commented on sentinel/src/schedule/replications.js:
```diff
@@ -37,11 +43,102 @@ function getSchedule(config) {
+      filter: doc => !doc._deleted && isTelemetryOrFeedback(doc._id),
+    })
+    .on('change', changes => {
+      return purgeDocs(sourceDb, changes.docs);
```
Only running this can become a problem if a user logs in on multiple devices: 
Example: 
1. User logs in on device A and generates feedback doc X.
2. device A syncs doc X upwards. 
3. user logs in on device B and generates feedback doc Y. 
4. device B syncs and uploads doc Y and downloads doc X. 
5. the scheduler runs the replication and purges both doc X and doc Y. 
6. device B syncs again and uploads doc X. 

At this point, doc X will never be purged, because it's already replicated to the conglomerate (_revs_diff will yield no missing revs, because the rev is the same, so it will be skipped in replication). 

Unsure of the correct approach. 
Should we just purge all existent and replicated feedback docs every time (like the oneTime but everyTime)
and just store a seq of the last "thing" we purged instead of starting from 0 when reading the changes feed?

garethbowen replied:
I think it's OK. It's unlikely to happen often and so there will only be a little bloat but not real impact.

dianabarsan commented on api/src/migrations/add-meta-validate-doc-update.js:
```diff
@@ -0,0 +1,29 @@
+const db = require('../db');
+const userDb = require('../services/user-db');
+
+const migradeDb = (dbName) => {
```
oops!

dianabarsan commented on api/src/migrations/add-meta-validate-doc-update.js:
```diff
@@ -0,0 +1,29 @@
+    .get('_design/medic-user')
+    .then(ddoc => {
+      ddoc.validate_doc_update = userDb.validateDocUpdate.toString();
+      return metaDb.put(ddoc);
```
Yes! Definitely! Thanks!

dianabarsan commented:
hey @garethbowen . I added all the changes you requested and also added my proposed idea of dealing with purging server-side, that accounts for users being logged in on multiple devices. I feel like it's an overall cleaner and clearer strategy that resembles what we do client-side. 
If you don't agree, I can easily revert the commits that add it. 

Thanks a lot for your time!
